README.md
# finto

[![Build Status](https://travis-ci.org/threadwaste/finto.svg?branch=master)](https://travis-ci.org/threadwaste/finto)

**finto (-a)** /'finto (-a)/ *agg* **1** posticcio; artificiàle

## Overview

finto is a web server that emulates EC2 instance profile roles on a workstation
through STS's assume role function. It was born as an experiment to ease local
interaction with AWS services in a deeply-federated, role-based environment.
finto ships with a basic API for moving between roles, and handles credentials
caching and expiration.

## Installation

At its simplest:

    go get github.com/threadwaste/finto

## Usage

    Usage of finto:
      -addr="169.254.169.254": bind to addr
      -config="/home/demo/.fintorc": location of config file
      -log="": log http to file
      -port=16925: listen on port

While running, finto provides credentials to EC2 instance profile providers.
This provider is last in the default provider chain of each SDK. For more
information, refer to the official documentation on [EC2 instance profile
roles](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html) and the [standardized credentials interface](https://blogs.aws.amazon.com/security/post/Tx3D6U6WSFGOK2H/A-New-and-Standardized-Way-to-Manage-Credentials-in-the-AWS-SDKs).

Below is sample output of finto serving credentials to the AWS CLI:

    $ aws s3 ls --debug
    <truncated>
    2016-01-03 11:52:01,895 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: env
    2016-01-03 11:52:01,895 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: assume-role
    2016-01-03 11:52:01,895 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: shared-credentials-file
    2016-01-03 11:52:01,896 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: config-file
    2016-01-03 11:52:01,896 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: ec2-credentials-file
    2016-01-03 11:52:01,896 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: boto-config
    2016-01-03 11:52:01,897 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: iam-role
    2016-01-03 11:52:01,902 - MainThread - botocore.vendored.requests.packages.urllib3.connectionpool - INFO - Starting new HTTP connection (1): 169.254.169.254
    2016-01-03 11:52:01,904 - MainThread - botocore.vendored.requests.packages.urllib3.connectionpool - DEBUG - "GET /latest/meta-data/iam/security-credentials/ HTTP/1.1" 200 5
    2016-01-03 11:52:02,259 - MainThread - botocore.vendored.requests.packages.urllib3.connectionpool - DEBUG - "GET /latest/meta-data/iam/security-credentials/example HTTP/1.1" 200 635
    2016-01-03 11:52:02,261 - MainThread - botocore.credentials - INFO - Found credentials from IAM Role: example
    <truncated>
    2016-01-03 11:52:03,282 - MainThread - botocore.hooks - DEBUG - Event after-call.s3.ListBuckets: calling handler <awscli.errorhandler.ErrorHandler object at 0x10483fc90>
    2016-01-03 11:52:03,282 - MainThread - awscli.errorhandler - DEBUG - HTTP Response Code: 200

finto also includes an API for bouncing between available roles. Helper
functions for bash and fish shells are available.

    $ curl 169.254.169.254/roles
    {"roles":["example","example2"]}
    $ curl 169.254.169.254/roles/example
    {"arn":"arn:aws:iam::123456789012:role/example","session_name":"finto-example"}
    $ curl 169.254.169.254/roles/example/credentials
    {
      "AccessKeyId": "<redacted>",
      "Code": "Success",
      "Expiration": "2016-01-03T19:40:30Z",
      "LastUpdated": "2015-07-07T23:06:33Z",
      "SecretAccessKey": "<redacted>",
      "Token": "<redacted>",
      "Type": "AWS-HMAC"
    }
    $ curl 169.254.169.254/latest/meta-data/iam/security-credentials/
    example
    $ curl -XPUT -d'{"alias":"example2"}' 169.254.169.254/roles
    {"active_role":"example2"}
    $ curl 169.254.169.254/latest/meta-data/iam/security-credentials/
    example2

## Configuration

finto uses a JSON configuration file to setup its credentials and the roles it
will serve. It currently uses a shared credentials provider only. Exluding the
credentials file or profile will use the defaults "~/.aws/credentials" and
"default," respectively.

    {
      "credentials": {
        "file": "/home/demo/.finto/credentials",
        "profile": "identity"
      },
      "roles": {
        "example": "arn:aws:iam::123456789012:role/example",
        "example2": "arn:aws:iam::123456789012:role/example2"
      }
      "default_role": "example",
    }

## Running

There are essentially two basic requirements for running finto:

  1. Routing the EC2 meta-data endpoint
  2. Using (or chaining to) the EC2 instance profile provider

The first can be achieved in several ways: interface aliasing, network
redirection, virtual machines, and so on. The wiki contains a couple of basic
examples.

The second is client-dependent. In the case of clients like the AWS CLI, the
user must clear a path to the EC2 instance profile provider. Multiple shared
credentials profiles can still be configured, and accessed with e.g. the
--profile option or AWS_DEFAULT_PROFILE environment variable.

## Development

After cloning the repository, running `make` will fetch and build
dependencies; run tests; and install the binary. During development, `make
testall` requires the following environment variables for its integration tests:

+ AWS_ACCESS_KEY_ID
+ AWS_SECRET_ACCESS_KEY
+ FINTO_VALID_ARN - an IAM role ARN that can be assumed
+ FINTO_INVALID_ARN - a false IAM role ARN that will fail

The target `test` can be used to skip the integration tests, and avoid this
setup.


CHANGELOG.md
## 0.0.1 (Jan 10, 2016)

+ Initial release


handlers.go
package finto

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/gorilla/mux"
)

// Contains application context.
type fintoContext struct {
	set          *RoleSet
	instanceRole string
}

func InitFintoContext(rs *RoleSet, defrole string) (fintoContext, error) {
	var fc = fintoContext{set: rs}
	err := fc.setInstanceRole(defrole)

	return fc, err
}

func (fc *fintoContext) setInstanceRole(role string) error {
	if _, err := fc.set.Role(role); err != nil {
		return err
	}

	fc.instanceRole = role
	return nil
}

// VarsHandlerFunc accepts mux route variables as an argument.
type VarsHandlerFunc func(http.ResponseWriter, *http.Request, map[string]string)

func (f VarsHandlerFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	f(w, r, vars)
}

// List available roles.
func rolesList(fc *fintoContext) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var roles []string

		if r.FormValue("status") == "active" {
			roles = []string{fc.instanceRole}
		} else {
			roles = fc.set.Roles()
		}

		jsonResponse(w, map[string][]string{"roles": roles})
	})
}

// Show a role's configuration.
func rolesShow(fc *fintoContext) http.Handler {
	return VarsHandlerFunc(func(w http.ResponseWriter, r *http.Request, vars map[string]string) {
		role, err := fc.set.Role(vars["alias"])
		if err != nil {
			errorResponse(w, err.Error(), http.StatusNotFound)
			return
		}

		jsonResponse(w, map[string]string{
			"arn":          role.Arn(),
			"session_name": role.SessionName(),
		})
	})
}

// Set role to be served as the instance profile role.
func rolesSetActive(fc *fintoContext) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		type activateRequest struct {
			Alias string `json:"alias"`
		}

		var req activateRequest

		decoder := json.NewDecoder(r.Body)
		if err := decoder.Decode(&req); err != nil {
			errorResponse(w, fmt.Sprint("failed to parse body: ", err),
				http.StatusBadRequest)
			return
		}

		if err := fc.setInstanceRole(req.Alias); err != nil {
			errorResponse(w, err.Error(), http.StatusBadRequest)
			return
		}

		jsonResponse(w, map[string]string{"active_role": fc.instanceRole})
	})
}

// Mock the EC2 security-credentials meta-data endpoint.
func mockProfile(fc *fintoContext) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(fc.instanceRole))
	})
}

// Mock the EC2 instance profile role meta-data endpoint.
func mockProfileCreds(fc *fintoContext) http.Handler {
	return VarsHandlerFunc(func(w http.ResponseWriter, r *http.Request, vars map[string]string) {
		role, err := fc.set.Role(vars["alias"])
		if err != nil {
			errorResponse(w, err.Error(), http.StatusNotFound)
			return
		}

		creds, err := role.Credentials()
		if err != nil {
			errorResponse(w, fmt.Sprint("failed to assume role: ", err),
				http.StatusInternalServerError)
			return
		}

		// There's technically no reason to pretty print here, but do so to
		// maintain parity in the mock service. Uses MarshalIndent as
		// Encoder.Encode does not offer a means to do so.
		b, err := json.MarshalIndent(map[string]string{
			"Code":            "Success",
			"LastUpdated":     "2015-07-07T23:06:33Z",
			"Type":            "AWS-HMAC",
			"AccessKeyId":     creds.AccessKeyId,
			"SecretAccessKey": creds.SecretAccessKey,
			"Token":           creds.SessionToken,
			"Expiration":      creds.Expiration.Format("2006-01-02T15:04:05Z"),
		}, "", "  ")

		if err != nil {
			errorResponse(w, fmt.Sprint("failed to render: ", err),
				http.StatusInternalServerError)
			return
		}

		w.Write(b)
	})
}

func jsonResponse(w http.ResponseWriter, body interface{}) {
	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	json.NewEncoder(w).Encode(body)
}

func errorResponse(w http.ResponseWriter, message string, code int) {
	w.WriteHeader(code)
	jsonResponse(w, map[string]string{"error": message})
}
